cityofaustin/atd-moped
# Switch to decoded URL params for tracking project list view state

## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/14588

This PR updates the current base64 encoded URL params in the project list view to contain less data and be url encoded. React Router handles the encoding since it uses [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) and that is the default behavior. 

This PR refactors the format of the data that we set and get from the URL, and it sets the stage to update the format of the `filterParameters` state within the `Filters` component too. For now, I created helpers to go between the previous and new formats as needed and document what the UI expects to be able to show the selected filters in the UI.

**What was previously stored in the URL params**
See the issue linked above

**What is stored with the new changes**
```
[
   { 
      "field":"type_name",
      "operator":"string_equals_case_sensitive",
      "value":"Fiber - New"
   },
   ...other filters
]
```

## Testing
**URL to test:** 

https://deploy-preview-1226--atd-moped-main.netlify.app/moped/projects/

**Steps to test:**
1. Go to the project list view and add advanced filters
2. As you add them, check that the list is filtering and that the url is storing new filters. [This site](https://meyerweb.com/eric/tools/dencoder/) always ends up coming up first when I look for a quick encode/decode tool so it could be helpful to see what is in the url.
3. Test a filter with lookup tables like **Type**
4. Test editing applied filters options (choose new field option, choose new operator option, or update search value)
5. Test deleting filters with the trash can icon
6. Test resetting the filters by clicking the reset button
7. Test making edits to the url string to make sure the app doesn't break when it can't parse the URL params
8. Test filtering, clicking on a project name link in the list view, and then clicking the "< ALL PROJECTS" link to go back to the list view. Your filters should still be present.

---
#### Ship list
